instrumentation/jdbi3/src/main/java/brave/jdbi3/Jdbi3BravePlugin.java

```diff
@@ -0,0 +1,173 @@
+            return;
+        }
+        jdbi.getConfig(SqlStatements.class)
+            .addContextListener(new StatementContextListener() {
+                @Override
+                public void contextCreated(final StatementContext ctx) {
+                    String spanName = ctx.describeJdbiStatementType();
+                    Span span = tracing.tracer().nextSpan().name(spanName);
+
+                    Instant start = ctx.getExecutionMoment();
+                    if (start != null) {
+                        span.start(start.toEpochMilli() * 1000);
+                    } else {
+                        span.start();
+                    }
+
+                    ctx.setTraceId(span.context().traceIdString());
+
+                    ctx.addCleanable(() -> {
+                        final SqlStatements stmtConfig = ctx.getConfig(SqlStatements.class);
+
+                        String remoteServiceName = getRemoteServiceName(ctx);
+                        if (remoteServiceName != null && !remoteServiceName.isEmpty()) {
+                            span.remoteServiceName(remoteServiceName);
+                        }
+
+                        span.kind(Span.Kind.CLIENT);
+
+                        final String renderedSql = ctx.getRenderedSql();
+                        if (renderedSql != null) {
+                            String truncated = renderedSql.substring(
+                                0,
+                                Math.min(renderedSql.length(), stmtConfig.getJfrSqlMaxLength())
+                            );
+                            span.tag("sql.query", truncated);
+                        }
+
+                        if (includeBindVariables) {
+                            String bindVariables = ctx.getBinding()
+                                .describe(stmtConfig.getJfrParamMaxLength());
+                            span.tag("sql.binding", bindVariables);
+                        }
+
+                        span.tag("sql.rows", Long.toString(ctx.getMappedRows()));
+
+                        if (ctx.getCompletionMoment() == null) {
+                            span.tag("error", "");
+                        }
+
+                        if (ctx.getCompletionMoment() != null) {
+                            span.finish(ctx.getCompletionMoment().toEpochMilli() * 1000);
+                        } else if (ctx.getExceptionMoment() != null) {
+                            span.finish(ctx.getExceptionMoment().toEpochMilli() * 1000);
+                        } else {
+                            span.finish();
+                        }
+                    });
+                }
+            });
+    }
+
+    static String getRemoteServiceName(StatementContext ctx) {
+        Connection connection = ctx.getConnection();
+        if (connection == null) {
+            return null;
+        }
+        URI uri;
+        try {
+            uri = new URI(connection.getMetaData().getURL().replaceAll("^jdbc:", ""));
```

Comment:
normally I would challenge you to make this similar to mysql code which is more efficient than regex. However, you've had a lot of back and forth so we can leave that for someone else to refine. Generally it is better to look at how we do things as often there is some performance mania going on, though not always.

https://github.com/openzipkin/brave/blob/ff4006230706a8ee357bc6ba18479faa2c3a86c2/instrumentation/mysql/src/main/java/brave/mysql/TracingStatementInterceptor.java#L76-L77

Reply:
Also added a simple cache, since a listener usually only ever sees one URL, and the remote service name for a given URL will always be the same.